### PR TITLE
Use Regex against Facts for setting default hostgroup

### PR DESCRIPTION
**Pull request by natewalck**
This change would allow the use of a regex against any given fact rather that just the hostname of the machine. 

The first match found  as listed in the config would be applied to the machine. As ruby 1.9.x retains the order of a hash read from config, this should be a reliable way of establishing the priority.

Eventually I would like to perhaps do compound comparisons (If factA = regex and FactB = regex), but this seems like a good start.

Open to any and all suggestions!

### CODE REVIEW

File: default_hostgroup.yaml.example [hunk 1]
```diff
@@ -2,20 +2,23 @@
 #
 # Must be placed in ~foreman/config/settings.plugins.d/
 #
-# :map:
-#   <hostgroup>:    <regex>
+# :facts_map:
+#   :<hostgroup>:
```
Comment (GregSutcliffe) [inline default_hostgroup.yaml.example]: Is there any reason for the Hostgroup to be a Symbol instead of a String? I could make it difficult to use slashes and/or spaces....

**Comment by natewalck** [comment]
I fixed the example to show slashes and spaces for hostgroup names work. I also confirmed this works on my dev foreman instance.

File: lib/default_hostgroup_managed_host_patch.rb [hunk 2]
```diff
@@ -36,18 +36,23 @@ def import_host_and_facts_with_match_hostgroup hostname, facts, certname = nil,
...
-      map.each do |hostgroup, regex|
-        unless valid_hostgroup?(hostgroup)
-          Rails.logger.error "DefaultHostgroupMatch: #{hostgroup} is not a valid hostgroup, skipping."
-          next
-        end
-        regex.gsub!(/(\A\/|\/\z)/, '')
-        if Regexp.new(regex).match(hostname)
-          new_hostgroup = Hostgroup.find_by_title(hostgroup)
-          break
+      catch (:foundmatch) do
```
Comment (GregSutcliffe) [inline lib/default_hostgroup_managed_host_patch.rb]: catch/throw is not style i've used much tbh, although there's nothing wrong with it. I'll have a play with it later and see what other options there might be :)

Comment (natewalck) [inline lib/default_hostgroup_managed_host_patch.rb]: If there is a more ruby-ish way to do it, I am open. The only goal is to exit the entire loop as soon as a match is found, otherwise everything goes in whatever the last match is (catchall in our example).

Comment (GregSutcliffe) [inline lib/default_hostgroup_managed_host_patch.rb]: Thinking about it, we can probably make this much more readable by breaking it down into some separate functions rather than one massive loop. As a bonus we can then use "return <foo>" in the function to break out of the loop once we have a match. Eg:

```
new_hostgroup = find_match(facts_map)

...

def find_match (facts_map)
  (psuedo code here) facts_map.each do |group, fact|
    return group if group_matches?(fact)
  end
end

def group_matches?(fact)
  (more psuedo code resulting in true/false)
end
```

what do you think? feels more readable that way, to me anyhow...

**Comment by GregSutcliffe** [comment]
I like the idea. We need to break that massive matcher routine down a bit, it's quite dense and hard to follow at the moment.

We'll also need tests (https://github.com/GregSutcliffe/foreman_default_hostgroup/blob/develop/test/unit/default_hostgroup_test.rb for existing examples) to excercise the new functionality - it's not too hard to extend, so I can probably help. You will need a dev setup if you wish to run the tests yourself though ;)

File: lib/default_hostgroup_managed_host_patch.rb (hunk 2 above)
Comment (natewalck) [inline lib/default_hostgroup_managed_host_patch.rb]: I like it. Does ruby pass the returned object up the chain?

Example:

```
      def find_match(facts_map)
        facts_map.each do |group, fact|
          return group if group_matches?(fact)
        end
      end
```

Would that pass 'group' all the way up and return it? Or do I need to return group again inside the find_match def itself?

Comment (natewalck) [inline lib/default_hostgroup_managed_host_patch.rb]: Nevermind! I figured it out in irb.

File: lib/default_hostgroup_managed_host_patch.rb [hunk 3]
```diff
@@ -36,19 +36,24 @@ def import_host_and_facts_with_match_hostgroup hostname, facts, certname = nil,
...
-      map = SETTINGS[:default_hostgroup][:map]
-      new_hostgroup = nil
+      facts_map = SETTINGS[:default_hostgroup][:facts_map]
+      new_hostgroup = find_match(facts_map)
 
-      map.each do |hostgroup, regex|
-        unless valid_hostgroup?(hostgroup)
-          Rails.logger.error "DefaultHostgroupMatch: #{hostgroup} is not a valid hostgroup, skipping."
-          next
+      def find_match(facts_map)
```
Comment (GregSutcliffe) [inline lib/default_hostgroup_managed_host_patch.rb]: `def ...` should not be within another `def ...` block - move them to the end of the file along with `valid_hostgroup?`

Comment (natewalck) [inline lib/default_hostgroup_managed_host_patch.rb]: Ah yes. This is what happens when you code with multiple interruptions every hour ;)

**Comment by natewalck** [comment]
Hey Greg, I fixed the functions and finally got around to testing them. They seem to be working well.

What is the next step?

**Comment by GregSutcliffe** [comment]
next step is me finding time to test it - something of a bottleneck :P

File: lib/default_hostgroup_managed_host_patch.rb [hunk 4]
```diff
@@ -60,6 +48,23 @@ def import_host_and_facts_with_match_hostgroup hostname, facts, certname = nil,
       return host, result
     end
 
+    def group_matches?(fact)
+      fact.each do |fact_name, fact_regex|
+        fact_regex.gsub!(/(\A\/|\/\z)/, '')
+        host_fact_value = host.facts_hash[fact_name]
```
Comment (GregSutcliffe) [inline lib/default_hostgroup_managed_host_patch.rb]: 'host' won't exist here, as it was defined in the calling method and not passed in to this one. You _could_ define the method as `group_matches?(fact, host)` and pass it in, but as these are all in the same class, I'd make host a class variable - change line 13 to 

```
@host, result = import_host_and_facts_without_match_hostgroup(hostname, facts, certname, proxy_id)
```

and use @host wherever you current use host

File: lib/default_hostgroup_managed_host_patch.rb [hunk 5]
```diff
@@ -22,42 +22,48 @@ def import_host_and_facts_with_match_hostgroup hostname, facts, certname = nil,
...
+        Rails.logger.info "Fact = #{fact_name}"
+        Rails.logger.info "Regex = #{fact_regex}"
+        return true if Regexp.new(fact_regex).match(host_fact_value)
+      end
+      return false
+    end
+
+    def find_match(facts_map)
+      facts_map.each do |group, facts|
+        return Hostgroup.find_by_title(group) if group_matches?(facts) and valid_hostgroup?(group)
```
Comment (GregSutcliffe) [inline lib/default_hostgroup_managed_host_patch.rb]: There's a few things here. First, names - lets go with group_name (so we can infer its just a string).

There's also no warning if we find no match, Perhaps we should add

```
Rails.logger.info "No match found for #{@host.name}"
return false
```

to the end of find_match?

**Comment by natewalck** [comment]
Down to one test fail:

https://gist.github.com/natewalck/29e758018a5c6723b7a9

**Comment by natewalck** [comment]
Based on that test, it seems like if the hostgroup is invalid, I should skip to the next result rather then returning false?

**Comment by natewalck** [comment]
Woot! Using continue instead of return false allowed it to process the rest of the items in the array instead of bailing on an invalid hostgroup.

All tests are passing now, please confirm I didn't break any of the tests in order to pass them (I don't think I did).

File: lib/default_hostgroup_managed_host_patch.rb [hunk 6]
```diff
@@ -22,42 +22,53 @@ def import_host_and_facts_with_match_hostgroup hostname, facts, certname = nil,
...
+        return true if Regexp.new(fact_regex).match(host_fact_value)
+      end
+      return false
+    end
+
+    def find_match(facts_map)
+      facts_map.each do |group_name, facts|
+        if group_matches?(facts) and valid_hostgroup?(group_name)
+          return Hostgroup.find_by_title(group_name)
+        else
```
Comment (GregSutcliffe) [inline lib/default_hostgroup_managed_host_patch.rb]: This else block is in the wrong place. As I said in my previous comment, it should be at the end of the method, what I meant was, outside the if:

```
def find_match(facts_map)
  facts_map.each do |group_name, facts|
    return Hostgroup.find_by_title(group_name) if group_matches?(facts) and valid_hostgroup?(group_name)
  end
  Rails.logger.info "No match ..."
  return false
end
```

File: lib/default_hostgroup_managed_host_patch.rb [hunk 7]
```diff
@@ -22,42 +22,53 @@ def import_host_and_facts_with_match_hostgroup hostname, facts, certname = nil,
...
-          return host, result
+          return @host, result
         end
       end
 
-      map = SETTINGS[:default_hostgroup][:map]
-      new_hostgroup = nil
+      facts_map = SETTINGS[:default_hostgroup][:facts_map]
+      new_hostgroup = find_match(facts_map)
+      puts new_hostgroup.inspect
```
Comment (GregSutcliffe) [inline lib/default_hostgroup_managed_host_patch.rb]: I don't think you meant to leave this here :D

**Comment by GregSutcliffe** [comment]
All current tests pass for me (even with the above change), nice work!

We now just need a couple of new tests to cover your new functions - the existing tests confirm the structure of the facts_map hash is fine, so thats good. I think we need tests for:
- multiple regexes for one group pick the right group, i.e. `"group" => { "fact1" => "foo", "fact2" => "bar"}` would match work on hosts having either fact
- multiple matches across groups would pick the right group. Given the nature of the return inside the loop, I think this would always be first match wins, but lets prove that behaviour in case we change it later

I think thats the only two, but I'll ruminate on it overnight.

File: lib/default_hostgroup_managed_host_patch.rb (hunk 6 above)
Comment (natewalck) [inline lib/default_hostgroup_managed_host_patch.rb]: What I am not understanding is how this evaluates. When you return the Hostgroup, it breaks the do loop, but why does it not return false every time as that comes after the do loop? Does that get skipped for some reason?

**Comment by natewalck** [comment]
For point #1, currently the first one it matches under the given hostgroup, it uses.

For instance:

```
"redhat base":
  "lsbdistid": "RedHat*"
  "osfamily": "RedHat"
```

It matches on the first one, doesn't even try the second one. I tend to like this behavior, but are you thinking we should do compound comparisons? Must match X fact _and_ Y fact?

Here is what the output looks like now:

```
Import facts for 'sinn1636.lan' completed. Added: 87, Updated: 0, Deleted 0 facts
Fact = lsbdistid
Regex = RedHat*
Fact = osfamily
Regex = RedHat
```

So it didn't get a match on the first fact within the config, but did get a match on the second fact.

File: lib/default_hostgroup_managed_host_patch.rb (hunk 6 above)
Comment (GregSutcliffe) [inline lib/default_hostgroup_managed_host_patch.rb]: `return` doesn't break the loop, it returns the specified object and exits the _method_. So the last return will not be sent if any match is made earlier.

**Comment by GregSutcliffe** [comment]
@natewalck no, I agree with you, I'm saying we need tests to prove thats what it does :) My example suggests that if the first one _doesn't_ match, the second will be used, and I'm fine with that behaviour, but we need to ensure the tests prove it so it isn't broken in the future.

**Comment by natewalck** [comment]
Ok, I've added the logic as described for exiting the loop and I've added a test that should work for exiting after first match.

For the test, I could feed it 5 items or just 2 as I have it and it shouldn't change the behavior.

**Comment by natewalck** [comment]
I confirmed the new test passed, but I still need a sanity test to make sure that the test does what I think it should do ;)

File: test/unit/default_hostgroup_test.rb [hunk 8]
```diff
@@ -48,7 +48,7 @@ def setup_hostgroup_match
 
   test "invalid hostgroup does nothing" do
```
Comment (GregSutcliffe) [inline test/unit/default_hostgroup_test.rb]: Can you rename this test to something like "invalid hostgroup name is ignored" ? I had to read the code twice to figure out what it was testing ;)

Comment (natewalck) [inline test/unit/default_hostgroup_test.rb]: Changed!

File: test/unit/default_hostgroup_test.rb [hunk 9]
```diff
@@ -57,9 +57,20 @@ def setup_hostgroup_match
...
+    setup_hostgroup_match
+    SETTINGS[:default_hostgroup][:facts_map] = { "First Group" => {'hostname' => '.*'}, "Second Group" => {'hostname' => '.*'} }
+
+    hostgroup = Hostgroup.create(:name => "First Group")
+    raw = parse_json_fixture('/facts.json')
+
+    assert Host.import_host_and_facts(raw['name'], raw['facts'])
+    assert_equal hostgroup, Host.find_by_name('sinn1636.lan').hostgroup
+  end
+
```
Comment (GregSutcliffe) [inline test/unit/default_hostgroup_test.rb:70]: We also need some tests for the multi-regex support. These hashes spring to mind:

```
invalid keys ignored: { "First Group" => { 'nosuchfact' => '.*', 'hostname' => '.*' }}
unmatched values ignored: { "First Group" => { 'hostname' => 'nosuchname', 'osfamily' => '.*' }}
```

**Comment by natewalck** [comment]
Tests added for the described scenario, passes:

```
# Running tests:

..........

Finished tests in 9.082091s, 1.1011 tests/s, 2.2021 assertions/s.
```

**Comment by GregSutcliffe** [comment]
Merged as fe7550e, thanks for the hard work @natewalck!
